sled-storage/src/manager.rs

```diff
@@ -315,191 +376,363 @@ impl StorageManager {
+        let maybe_ledger = Ledger::<OmicronPhysicalDisksConfig>::new(
+            &log,
+            ledger_paths.clone(),
+        )
+        .await;
+
+        match maybe_ledger {
+            Some(ledger) => {
+                info!(self.log, "Ledger of physical disks exists");
+                return Some(ledger);
+            }
+            None => {
+                info!(self.log, "No ledger of physical disks exists");
+                return None;
+            }
         }
+    }
 
-        match Disk::new(&self.log, raw_disk.clone(), Some(&self.key_requester))
-            .await
-        {
-            Ok(disk) => self.resources.insert_disk(disk),
-            Err(err @ DiskError::Dataset(DatasetError::KeyManager(_))) => {
+    async fn key_manager_ready(&mut self) -> Result<(), Error> {
+        // Unconditionally set the state to "synchronization needed",
+        // to force us to try to asynchronously ensure that disks are ready.
+        self.state = StorageManagerState::SynchronizationNeeded;
+
+        // Now that we're actually able to unpack U.2s, attempt to load the
+        // set of disks which we previously stored in the ledger, if one
+        // existed.
+        let ledger = self.load_ledger().await;
+        if let Some(ledger) = ledger {
+            info!(self.log, "Setting StorageResources state to match ledger");
+
+            // Identify which disks should be managed by the control
+            // plane, and adopt all requested disks into the control plane
+            // in a background task (see: [Self::manage_disks]).
+            self.resources.set_config(&ledger.data().disks);
+        } else {
+            info!(self.log, "KeyManager ready, but no ledger detected");
+            if !self.resources.get_config().await.is_empty() {
+                // This case seems unlikely -- we'd need to somehow have no
+                // ledger, but have been told to store the configuration
+                // regardless -- but it seems safe to bail out if we have
+                // ANY configuration stored, for any reason.
                 warn!(
                     self.log,
-                    "Transient error: {err}: queuing disk";
-                    "disk_id" => ?raw_disk.identity()
+                    "Some configuration already exists, without ledger"
                 );
-                self.queued_u2_drives.insert(raw_disk);
-                self.state = StorageManagerState::QueueingDisks;
-                Ok(AddDiskResult::DiskQueued)
+                return Ok(());
             }
-            Err(err) => {
-                error!(
-                    self.log,
-                    "Persistent error:not queueing disk";
-                    "err" => ?err,
-                    "disk_id" => ?raw_disk.identity()
+
+            info!(self.log, "No configuration exists");
+
+            // NOTE: What follows is an exceptional case: one where we have
```

Comment:
This looks great.

Unfortunately, I think there might be a race condition here, which would be problematic for existing non-ledger systems. It's possible that the key-manager will become ready before all the disks are detected on a cold boot. Unfortunately, then an existing disk with a zpool will not become a member of the control plane disks and will not become managed. When `detected_raw_disk` is called we should check if we are in the ledgerless scenario, and then do the same thing we do here to see if there is a valid zpool. If so we should add it to the config.

We should also add a test for this. I marked the spot below where I think it can slot into an existing test.

Reply:
Good catch, I've added a test for this in b65b4c124d1dc80e4830c866ec3b2d214aaba433, and (I believe) fixed this race by "re-loading the config" whenever we add a U.2 without having a ledger.